pkg/kubelet/cri/remote/remote_runtime.go

```diff
@@ -1149,3 +1150,82 @@ func (r *remoteRuntimeService) ReopenContainerLog(containerID string) (err error
 	klog.V(10).InfoS("[RemoteRuntimeService] ReopenContainerLog Response", "containerID", containerID)
 	return nil
 }
+
+func (r *remoteRuntimeService) CheckpointContainer(containerID, checkpointDir string) error {
+	klog.V(10).InfoS(
+		"[RemoteRuntimeService] CheckpointContainer",
+		"containerID",
+		containerID,
+		"directory",
+		checkpointDir,
+	)
+	ctx, cancel := getContextWithTimeout(r.timeout)
+	defer cancel()
+
+	api, err := func() (string, error) {
+		if r.useV1API() {
+			_, err := r.runtimeClient.CheckpointContainer(
+				ctx,
+				// The CRI implementation reads the ID and checkpoints
+				// a container if the ID represents a container or it
+				// checkpoints a Pod if the ID represents a Pod.
+				&runtimeapi.CheckpointContainerRequest{
+					Id: containerID,
+					Options: &runtimeapi.CheckpointContainerOptions{
+						CommonOptions: &runtimeapi.CheckpointRestoreOptions{
+							// Currently the archive is not yet compressed, but
+							// the CRI API already contains the necessary field
+							// to tell the CRI implementation which compression
+							// to use.
+							ArchiveLocation: filepath.Join(checkpointDir, containerID+".tar"),
+						},
+						// For the first try of checkpoint we want the container
+						// to not stop after checkpointing.
+						LeaveRunning: true,
+					},
+				},
+			)
+			return "V1", err
+		}
+		_, err := r.runtimeClientV1alpha2.CheckpointContainer(
+			ctx,
+			// The CRI implementation reads the ID and checkpoints
+			// a container if the ID represents a container or it
+			// checkpoints a Pod if the ID represents a Pod.
+			&runtimeapiV1alpha2.CheckpointContainerRequest{
```

Comment:
> at some point we should stop adding new apis to old versions of the api.. :-)

I am just following the existing code and it currently seems to support v1 and v1alpha2.

> Maybe add a check if returned error v1 and v1alpha2 responds that the runtime does not include this new CheckpointContainer service.. Or is the grpc error clean enough output?

I am not sure I understand this comment correctly. Why should there be a check? Can you clarify what you meant.

Reply:
> I am just following the existing code and it currently seems to support v1 and v1alpha2.

We should now be treating v1alpha2 as frozen. Confirm with @mrunalp. This, at least because, older versions of container runtimes implementing the v1alpha2 api will not have your new service api.

> > Maybe add a check if returned error v1 and v1alpha2 responds that the runtime does not include this new CheckpointContainer service.. Or is the grpc error clean enough output?
> 
> I am not sure I understand this comment correctly. Why should there be a check? Can you clarify what you meant.

Older versions of container runtimes that do not have the new API will cause you to get a GRPC error here.

Reply:
I just tried it without the checkpoint CRI call implemented and I get a 500 with following text: "checkpointing of counter failed (container counter not found)"

That is not correct. I will look into getting the correct error message to the user. Good point.

Reply:
@mrunalp Should I remove everything concerning v1alpha2 from this PR?

Reply:
Yes, we shouldn't add features to v1alpha2.